incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -38,25 +59,306 @@ type ObjectReconcilerNew struct {
+		//ns := inst.GetNamespace()
+		//inst = object.Canonical(so)
+		//inst.SetNamespace(ns)
+		metadata.SetLabel(inst, api.LabelInheritedFrom, so.GetNamespace())
+		return overwriting, so
+	}
+
+	return unchanging, nil
+}
+
+// syncSource syncs the copy in the forest with the current source object. If there's a change,
+// enqueue all the descendants to propagate the new source.
+func (r *ObjectReconcilerNew) syncSource(ctx context.Context, log logr.Logger, src *unstructured.Unstructured) {
+	// syncPropagated doesn't check isExcluded because the propagation won't happen if it's excluded.
+	if r.isExcluded(log, src) {
+		return
+	}
+	sns := src.GetNamespace()
+	n := src.GetName()
+	origCopy := r.Forest.Get(sns).GetOriginalObject(r.GVK, n)
+
+	// Early exit if the source object exists and remains unchanged.
+	if origCopy != nil && reflect.DeepEqual(object.Canonical(src), object.Canonical(origCopy)) {
+		log.Info("Unchanged Source")
+		return
+	}
+
+	// Update or create a copy of the source object in the forest
+	r.Forest.Get(sns).SetOriginalObject(src.DeepCopy())
+
+	// Enqueue all the descendant copies
+	r.enqueueDescendants(ctx, log, src)
 }
 
 // SyncNamespace can be called manually by the HierarchyReconciler when the hierarchy changes.
+// It enqueues all the current objects in the namespace and local copies of the original objects
+// in the ancestors.
 func (r *ObjectReconcilerNew) SyncNamespace(ctx context.Context, log logr.Logger, ns string) error {
-	log.Info("SyncNamespace placeholder")
+	log = log.WithValues("gvk", r.GVK)
+	ul := &unstructured.UnstructuredList{}
+	ul.SetGroupVersionKind(r.GVK)
+	if err := r.List(ctx, ul, client.InNamespace(ns)); err != nil {
+		log.Error(err, "Couldn't list objects")
+		return err
+	}
+
+	// Enqueue all the current objects in the namespace because some of them may have been deleted.
+	for _, inst := range ul.Items {
+		if r.isExcluded(log, &inst) {
+			continue
+		}
+		co := object.Canonical(&inst)
+		co.SetNamespace(inst.GetNamespace())
+		log.Info("Enqueuing existing object for reconciliation", "affected", co.GetName())
+		r.Affected <- event.GenericEvent{Meta: co}
+	}
+
+	// Enqueue local copies of the originals in the ancestors to catch any new or changed objects.
+	r.enqueuePropagatedObjects(ctx, log, ns)
+
 	return nil
 }
 
-func (r *ObjectReconcilerNew) SetupWithManager(mgr ctrl.Manager) error {
+func (r *ObjectReconcilerNew) enqueueDescendants(ctx context.Context, log logr.Logger, src *unstructured.Unstructured) {
+	sns := src.GetNamespace()
+	dns := r.Forest.Get(sns).DescendantNames()
+	for _, ns := range dns {
+		dc := object.Canonical(src)
+		dc.SetNamespace(ns)
+		log.Info("Enqueuing descendant copy", "affected", ns+"/"+src.GetName(), "reason", "The source changed")
+		r.Affected <- event.GenericEvent{Meta: dc}
+	}
+}
+
+func (r *ObjectReconcilerNew) enqueuePropagatedObjects(ctx context.Context, log logr.Logger, ns string) {
+	r.Forest.Lock()
```

Comment:
I assume this is the reason you put this into its own function but not the other loop in `SyncNamespace`? Maybe worth commenting that.